Data/Primitive/Array.hs

```diff
@@ -520,18 +524,34 @@ traverseArray f = \ !ary ->
 {-# INLINE [1] traverseArray #-}
 
 {-# RULES
-"traverse/ST" forall (f :: a -> ST s b). traverseArray f =
-   traverseArrayP f
-"traverse/IO" forall (f :: a -> IO b). traverseArray f =
-   traverseArrayP f
+"toWonk" [~1] traverseArray
+  = traverseArrayWonk (WonkP (\(Dict :: Dict (PrimMonad f)) (g :: a -> f b) -> traverseArrayP g))
+"wonkIO" forall (w :: WonkP IO f).
+  traverseArrayWonk w = runWonkP w Dict
+"wonkST" forall (w :: WonkP (ST s) f).
+  traverseArrayWonk w = runWonkP w Dict
+"wonkMaybeT" forall (w :: WonkP (MaybeT m) f).
+  traverseArrayWonk w = traverseArrayWonk (WonkP (\(Dict :: Dict (PrimMonad m)) -> runWonkP w Dict))
+"wonkStateT" forall (w :: WonkP (StateT s m) f).
+  traverseArrayWonk w = traverseArrayWonk (WonkP (\(Dict :: Dict (PrimMonad m)) -> runWonkP w Dict))
+"wonkIdentityT" forall (w :: WonkP (IdentityT m) f).
+  traverseArrayWonk w = traverseArrayWonk (WonkP (\(Dict :: Dict (PrimMonad m)) -> runWonkP w Dict))
  #-}
-#if MIN_VERSION_base(4,8,0)
-{-# RULES
-"traverse/Id" forall (f :: a -> Identity b). traverseArray f =
-   (coerce :: (Array a -> Array (Identity b))
-           -> Array a -> Identity (Array b)) (fmap f)
- #-}
-#endif
+
+data Dict c where
+  Dict :: c => Dict c
+
+newtype WonkP m f = WonkP
+  { runWonkP :: forall a b. Dict (PrimMonad m) -> (a -> f b) -> Array a -> f (Array b) }
+
+traverseArrayWonk
+  :: Applicative f
+  => WonkP m f
+  -> (a -> f b)
+  -> Array a
+  -> f (Array b)
+traverseArrayWonk _ f = traverseArray f
+{-# INLINE [0] traverseArrayWonk #-}
```

Comment:
Note: if it would help, we could add another wonky parameter to this function to handle `Maybe`, `Either`, `Identity`, and the like. I just haven't yet been able to figure out how that part should work at all.